lib/modem_battery/modem_battery.c

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdio.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <modem/modem_battery.h>
+#include <modem/at_monitor.h>
+#include <modem/lte_lc.h>
+#include <modem/nrf_modem_lib.h>
+#include <nrf_modem_at.h>
+
+#define AT_XVBATLVL_SET_ENABLE	"AT%%XVBATLVL=1"
+#define AT_XVBATLVL_SET_DISABLE	"AT%%XVBATLVL=0"
+#define AT_XVBATLOWLVL_SET      "AT%%XVBATLOWLVL=%d"
+#define AT_XVBAT_SET            "AT%XVBAT"
+#define AT_XVBATLOWLVL_READ     "AT%XVBATLOWLVL?"
+#define AT_XPOFWARN_SET_ENABLE  "AT%%XPOFWARN=1,%d"
+#define AT_XPOFWARN_SET_DISABLE "AT%%XPOFWARN=0"
+
+LOG_MODULE_REGISTER(modem_battery, CONFIG_MODEM_BATTERY_LOG_LEVEL);
+
+AT_MONITOR(xvbatlowlvl_mon, "%XVBATLOWLVL", at_handler_xvbatlowlvl);
+
+LTE_LC_ON_CFUN(modem_battery_cfun_hook, on_cfun, NULL);
+
+static modem_battery_low_level_handler_t xvbatlowlvl_handler;
+static modem_battery_pofwarn_handler_t pofwarn_handler;
+
+static void pofwarn_event_handler(const struct lte_lc_evt *const evt)
+{
+	if (evt->type == LTE_LC_EVT_MODEM_EVENT &&
+	    evt->modem_evt == LTE_LC_MODEM_EVT_BATTERY_LOW) {
+		if (pofwarn_handler == NULL) {
+			return;
+		}
+
+		pofwarn_handler();
+	}
+}
+
+static void at_handler_xvbatlowlvl(const char *response)
+{
+	int ret;
+	int battery_voltage;
+
+	__ASSERT_NO_MSG(response != NULL);
+
+	LOG_DBG("%%XVBATLOWLVL notification: %s", response);
+
+	ret = sscanf(response, "%%XVBATLOWLVL: %d", &battery_voltage);
+	/* Match exactly once. */
+	if (ret != 1) {
+		LOG_ERR("Failed to read battery low level notification, error: %d", ret);
+		return;
+	}
+
+	if (xvbatlowlvl_handler == NULL) {
+		return;
+	}
+
+	xvbatlowlvl_handler(battery_voltage);
+}
+
+int modem_battery_low_level_handler_set(modem_battery_low_level_handler_t handler)
+{
+	if (handler == NULL) {
+		LOG_ERR("Invalid handler (handler=0x%08X)", (uint32_t)handler);
+		return -EINVAL;
+	}
+
+	xvbatlowlvl_handler = handler;
+
+	return 0;
+}
+
+int modem_battery_pofwarn_handler_set(modem_battery_pofwarn_handler_t handler)
+{
+	if (handler == NULL) {
+		LOG_ERR("Invalid handler (handler=0x%08X)", (uint32_t)handler);
+		return -EINVAL;
+	}
+
+	pofwarn_handler = handler;
+
+	lte_lc_register_handler(pofwarn_event_handler);
+
+	return 0;
+}
+
+int modem_battery_pofwarn_enable(int voltage)
+{
+	int err;
+
+	err = lte_lc_modem_events_enable();
+	if (err) {
+		LOG_ERR("Failed to enable modem events, error: %d", err);
+		return -EFAULT;
+	}
+
+	err = nrf_modem_at_printf(AT_XPOFWARN_SET_ENABLE, voltage);
+	if (err) {
+		return -EFAULT;
+	}
+
+	return 0;
+}
+
+int modem_battery_pofwarn_disable(void)
+{
+	int err;
+
+	err = lte_lc_modem_events_disable();
+	if (err) {
+		LOG_ERR("Failed to disable modem events, error: %d", err);
+		return -EFAULT;
+	}
```

Comment:
This will disable the LTE LC events for the whole application. Consider registering an AT monitor for MDEV here, and pausing that instead.